labs/SAW/SAW.md

```diff
@@ -0,0 +1,987 @@
+    self.assertIs( arrayAddNewVar05_result.is_success(), True)
+    self.assertIs( arrayAddNewVar10_result.is_success(), True)
+    self.assertIs(   array5AddAlias_result.is_success(), True)
+```
+
+### Preconditions and Indices
+
+//SHOW SOMETHING FAILING BECAUSE LACK OF PRECONDITION. For this could have a function that takes in an arbitrary sized array, but the function that uses it only has an array of size 5. Then explain how the SAW, when analyzing the first function in isolation, has no way of knowing the input in use cases actually has fixed size 5 so you need a precondition
+
+A common situation where one might use `precondition_f` is when an
+integer is passed into a function and is used to index into an array:
+ 
+### Explicit Arrays
+
+Another way to initialize arrays is through the `array` command. For
+example, suppose I have the following C function (taken from
+[here](https://github.com/GaloisInc/saw-script/blob/337ca6c9edff3bcbcd6924289471abd5ee714c82/saw-remote-api/python/tests/saw/test-files/llvm_array_swap.c)):
+
+```C
+#include <stdint.h>
+
+void array_swap(uint32_t a[2]) {
+    uint32_t tmp = a[0];
+    a[0] = a[1];
+    a[1] = tmp;
+}
+```
+
+Here's a contract that doesn't even require Cryptol (taken from
+[here](https://github.com/GaloisInc/saw-script/blob/337ca6c9edff3bcbcd6924289471abd5ee714c82/saw-remote-api/python/tests/saw/test_llvm_array_swap.py)):
+
+```python
+class ArraySwapContract(Contract):
+    def specification(self):
+        a0 = self.fresh_var(i32, "a0")
+        a1 = self.fresh_var(i32, "a1")
+        a  = self.alloc(array_ty(2, i32), points_to=array(a0, a1))
+
+        self.execute_func(a)
+
+        self.points_to(a[0], a1)
+        self.points_to(a[1], a0)
+        self.returns(void)
+```
+
+Explicit arrays can be useful when you want to assert a condition on a
+particular element of the array. Of course, a drawback is you have to
+initialize every member of the array. However, this can be problematic
+for large arrays. Ideally, one would just have to declare the array
+implicitly as before and then pass this array to a Cryptol
+specification for postconditions as was done in the previous examples.
+
+
+# Structs 
+
+//Dj help
+
+Consider the following struct:
+
+```C
+typedef struct {
+  uint32_t *vector;
+  uint32_t size;
+} 
+```
+
+### Struct Initialization and Explicit Structs
+
+### Structs in Cryptol
+
+# Using Gained Knowledge
+
+## Assumptions and Lemmas
+
+Sometimes one might want to black-box certain parts of the verification
+such as functions coming from libraries, especially if they have
+already been formally verified. For example, the height of
+[height-balanced binary
+trees](https://en.wikipedia.org/wiki/Self-balancing_binary_search_tree)
+on n nodes is given by the ceiling of `log(n)`. In C we might use the
+`math` library to write the helper function.
+
+```C
+#include <math.h>
+
+uint64_t log2i(uint64_t i) {
+  return ceil(log2(i));

```

Comment:
@smithnormform `ceil(log2(i))` in C is not equal to Cryptol's `lg2` for values `1 << j + 1 where j >= 49`. To stick w/ a builtin function that needs an assumption, you could instead use:

```
int __builtin_clzll (unsigned long long)
```
More info here: https://en.wikipedia.org/wiki/Find_first_set#CLZ

Reply:
I guess you could instead simply mention the disparity here and add a precondition to the SAW contract.

Reply:
Updated in SAW.md now!